### PR TITLE
#1380 Issue: extend the source with new streams

### LINKED ISSUE #1380: Expand Sendgrid source support
## Tell us about the problem you're trying to solve
Support the following entities from the sendgrid API. 

Contacts 
Global Suppressions
Suppression Groups
Suppression Group Members
Lists
Lists Recipients
Segments
Segment Recipients
Campaigns
Templates
Invalid Emails
Bounces
Blocks
Spam Reports
Campaigns

For this issue, we do not need to support incremental sync on these streams. We'll support incremental sync in the following ticket. (making 2 separate tickets just to break up the work)

Please make sure that: 
1. In the standard tests for this source, _every_ stream is tested (and that every stream outputs records) unless there is a rate limiting issue
2. you create fake data for every stream -- whoever is working on this issue, feel free to reach out to me if you need support

### PR DESCRIPTION
**Pull request by yevhenii-ldv**
## What
*Add new streams:*
* *Segments*
* *Templates*
* *Global Suppressions*
* *Suppression Groups*
* *Suppression Group Members*
* *Blocks*
* *Bounces*
* *Invalid Emails*
* *Spam Reports*

## Pre-merge Checklist
- [ ] *Run integration tests*
- [ ] *Publish Docker images*

## Recommended reading order
1. `test.java`
1. `component.ts`
1. the rest

### CODE REVIEW

**Comment by yevhenii-ldv** [comment]
I filled Sendgrid account with data for all streams except:
```
Bounces
Invalid Emails
Spam Reports
```

All of these streams are related to mailing, and I was unable to create data for them.

**Review by sherifnada — CHANGES_REQUESTED** [review]
Some comments: 
1. Why do we sweep non 200 responses under the rug? This is pretty dangerous -- it's a silent failure
2. This method of doing pagination looks good 👍 nicely done. Though metadata pagination can be made DRY

File: airbyte-integrations/connectors/source-sendgrid/source_sendgrid/client.py [hunk 1]
```diff
@@ -22,47 +22,109 @@
...
+                yield from stream_data
+
+                if len(stream_data) < PAGE_SIZE:
+                    return
+                offset += PAGE_SIZE
+        elif pagination_type == "metadata":
+            response = self._request(f"{self.BASE_URL}{stream_url}", params={"page_size": PAGE_SIZE})
+            if response.status_code != status_codes.OK:
+                return
+            stream_data = (response.json()[key] if key else response.json()) or []
```
[the remark this answers is not in the document]
Comment (yevhenii-ldv) [reply airbyte-integrations/connectors/source-sendgrid/source_sendgrid/client.py:83]: If the account does not have data for the `stats_automations` stream, then in the response we get not an empty array `[ ]`, but `None`.

File: airbyte-integrations/connectors/source-sendgrid/source_sendgrid/client.py [hunk 2]
```diff
@@ -22,47 +22,109 @@
...
+    def _request(self, url: str, **kwargs) -> Response:
+        response = requests.get(url, headers=self._headers, **kwargs)
+        if response.status_code == status_codes.TOO_MANY_REQUESTS:
+            raise TooManyRequests()
+        return response
+
+    def _paginator(self, stream_url: str, pagination_type: str = None, key: str = None) -> Generator[dict, None, None]:
+        if not pagination_type:
+            response = self._request(f"{self.BASE_URL}{stream_url}")
+            if response.status_code == status_codes.OK:
```
[the remark this answers is not in the document]
Comment (yevhenii-ldv) [reply airbyte-integrations/connectors/source-sendgrid/source_sendgrid/client.py]: fixed it

File: airbyte-integrations/connectors/source-sendgrid/source_sendgrid/client.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sherifnada) [reply airbyte-integrations/connectors/source-sendgrid/source_sendgrid/client.py]: could you add this as a comment to the code?

File: airbyte-integrations/connectors/source-sendgrid/source_sendgrid/client.py [hunk 3]
```diff
@@ -22,47 +22,113 @@
...
+
+    def _read_data(self, stream_url: str, pagination: Pagination, key: str = None) -> Generator[dict, None, None]:
+        if pagination == Pagination.Non_pagination:
+            response = self._request(f"{self.BASE_URL}{stream_url}")
+            yield from response.json()[key] if key else response.json()
+        elif pagination == Pagination.Offset:
+            offset = 0
+            while True:
+                response = self._request(f"{self.BASE_URL}{stream_url}?limit={PAGE_SIZE}&offset={offset}")
+                stream_data = response.json()[key] if key else response.json()
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-sendgrid/source_sendgrid/client.py:81]: should this also contain a `or []`? I guess I'm unsure when we should expect that vs. not. Could you add a comment below `def _read_data` that explains this?

**Review by sherifnada — APPROVED** [review]
LGTM with one last comment. Please ping me when you've addressed and I'll publish.

File: airbyte-integrations/connectors/source-sendgrid/source_sendgrid/client.py [hunk 4]
```diff
@@ -22,47 +22,113 @@
...
+                yield from stream_data
+
+                if len(stream_data) < PAGE_SIZE:
+                    return
+                offset += PAGE_SIZE
+        elif pagination == Pagination.Metadata:
+            next_page = True
+            while next_page:
+                response = self._request(f"{self.BASE_URL}{stream_url}", params={"page_size": PAGE_SIZE})
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-sendgrid/source_sendgrid/client.py:90]: We don't see to be using `next_page` at all here...I think this isn't paging correctly

**Review by sherifnada — CHANGES_REQUESTED** [review]
I think there is a bug with paging

File: airbyte-integrations/connectors/source-sendgrid/source_sendgrid/client.py (hunk 3 above)
Comment (yevhenii-ldv) [reply airbyte-integrations/connectors/source-sendgrid/source_sendgrid/client.py:81]: added comment

File: airbyte-integrations/connectors/source-sendgrid/source_sendgrid/client.py (hunk 4 above)
Comment (yevhenii-ldv) [reply airbyte-integrations/connectors/source-sendgrid/source_sendgrid/client.py]: And the truth is, I don't know how I could have missed this, it turns out to be an endless loop. Already fixed this bug.

**Pull request merged**